jmclient/jmclient/wallet_service.py

```diff
@@ -0,0 +1,362 @@
+            # Python, since they are dicts. However their keyset is deterministic
+            # so it is sufficient to convert these dicts to tuples with fixed
+            # ordering, thus it can be used as a key into the self.callbacks
+            # dicts. (This is needed because txid is not always available
+            # at the time of callback registration).
+            possible_keys = [txid, tuple(
+                    (x["script"], x["value"]) for x in txd["outs"])]
+
+            # note that len(added_utxos) > 0 is not a sufficient condition for
+            # the tx being new, since wallet.add_new_utxos will happily re-add
+            # a utxo that already exists; but this does not cause re-firing
+            # of callbacks since we in these cases delete the callback after being
+            # called once.
+            # Note also that it's entirely possible that there are only removals,
+            # not additions, to the utxo set e.g. in tumbler sweeps or direct
+            # send payments.
+            if len(added_utxos) > 0 or len(removed_utxos) > 0:
+                if confs == 0:
+                    for k in possible_keys:
+                        if k in self.callbacks["unconfirmed"]:
+                            for f in self.callbacks["unconfirmed"][k]:
+                                # True implies success, implies removal:
+                                if f(txd, txid):
+                                    self.callbacks["unconfirmed"][k].remove(f)
+                                    # keep monitoring for conf > 0:
+                                    self.active_txids.append(txid)
+                elif confs > 0:
+                    for k in possible_keys:
+                        if k in self.callbacks["confirmed"]:
+                            for f in self.callbacks["confirmed"][k]:
+                                if f(txd, txid, confs):
+                                    self.callbacks["confirmed"][k].remove(f)
+                                    if txid in self.active_txids:
+                                        self.active_txids.remove(txid)
+
+    def check_callback_called(self, txinfo, callback, cbtype, msg):
+        """ Intended to be a deferred Task to be scheduled some
+        set time after the callback was registered.
+        """
+        assert cbtype in ["all", "unconfirmed", "confirmed"]
+        if txinfo in self.callbacks[cbtype]:
+            if callback in self.callbacks[cbtype][txinfo]:
+                # the callback was not called, drop it and warn
+                self.callbacks[cbtype][txinfo].remove(callback)
+                jlog.info("Timed out: " + msg)
+                # if callback is not in the list, it was already
+                # processed and so do nothing.
+
+    """ The following functions mostly are pure pass through
+    but are replicated here from the underlying wallet to
+    preserve encapsulation.
+    """
+
+    def get_txtype(self):
```

Comment:
All those wallet pass-through methods in `WalletService` could be replaced with a single `__getattr__ ` implementation, that simply calls the wallet whenever an unknown method is called on the `WalletService`. This will make the interface much easier to maintain.

Reply:
Oh cool idea, thanks.

Reply:
Done in https://github.com/JoinMarket-Org/joinmarket-clientserver/pull/359/commits/248dbbad63bd27862e7b3b9354468db13d09aec3